config: lower outgoing htlc broadcast delta

joostjager opened the pull request:
Builds on top of #2879.

We create outgoing htlcs and accept incoming htlcs within the go-to-chain window, risking triggering a force close. This PR updates the block deltas to make sure that we don't trigger a force close ourselves.

The main change is lowering the broadcast delta for outgoing htlcs. It can be lower than the delta for incoming htlcs, as we don't risk not being able to settle with the preimage if the tx(es) don't confirm in time.

The spec even suggests a delta of -1 for outgoing htlcs (https://github.com/lightningnetwork/lightning-rfc/blob/master/02-peer-protocol.md#cltv_expiry_delta-selection). Meaning that we only go to chain in the block after the block where the htlc expires.

This PR doesn't lower it to -1, but uses 0 instead to prevent confusion around the negative number. This means that we go to chain when an outgoing htlc expires. 

`OutgoingCltvRejectDelta` is a few blocks more than `OutgoingBroadcastDelta`. This is to prevent us forwarding an htlc that would bring us inside the broadcast window. We would probably receive a settle or fail quickly, but if a new block arrives in the mean time, we'd force close the channel.

`FinalCltvRejectDelta` could have been equal to `IncomingBroadcastDelta`, because as an exit hop we can immediately decide what to do with the incoming htlc.

This PR does increase `FinalCltvRejectDelta` a bit because for hodl invoices we do not decide immediately. Eventually, for hodl invoices, there will be automatic cancellation to prevent channel closure. This automatic cancellation should trigger some blocks before we go to chain. We can probably use the same value `FinalCltvRejectDelta` and the delta at which we cancel back hodl invoice htlcs.

joostjager commented:
Comments from @halseth 
- Add unit test to assert spec compliance of delta constant
- Reword comment on `IncomingBroadcastDelta` in `chain_arbitrator.go`

joostjager commented:
Reminder: Also need to check the outgoing expiry when we are the sender of the payment. In that case, `HtlcSatisfyPolicy` isn't called.

cfromknecht commented:
```
# github.com/lightningnetwork/lnd [github.com/lightningnetwork/lnd.test]
./lnd_test.go:9473:33: constant -4 overflows uint32
```

joostjager commented:
@cfromknecht ptal

Implemented policy check for local htlcs.

Didn't implement unit test to assert spec compliance. It doesn't look very useful to me to just check that a constant has a value. Instead I added extensive comments to the config constants which should future devs make think about what these values mean.

cfromknecht commented on htlcswitch/switch_test.go:
```diff
@@ -1382,8 +1397,9 @@ func TestSkipIneligibleLinksLocalForward(t *testing.T) {
 
 	aliceChannelLink := newMockChannelLink(
-		s, chanID1, aliceChanID, alicePeer, false,
+		s, chanID1, aliceChanID, alicePeer, eligable,
```
s/eligable/eligible

joostjager commented:
@cfromknecht ptal

halseth commented on config.go:
```diff
@@ -68,11 +68,50 @@ const (
+	defaultOutgoingBroadcastDelta = 0
+
+	// defaultOutgoingCltvRejectDelta defines the number of blocks before
+	// the expiry of an outgoing htlc at which we don't want to it offer to
```
s/it offer/offer it

halseth commented on htlcswitch/link.go:
```diff
@@ -2195,6 +2165,36 @@ func (l *channelLink) HtlcSatifiesPolicy(payHash [32]byte,
 		return &lnwire.FailExpiryTooFar{}
 	}
 
+	// Next, using the amount of the incoming HTLC, we'll calculate the
```
commit message should explain the rationale behind this change.

joostjager replied:
added

halseth commented on config.go:
```diff
@@ -68,11 +68,50 @@ const (
-
-	defaultOutgoingBroadcastDelta  = 10
-	defaultOutgoingCltvRejectDelta = 0
+	// defaultIncomingBroadcastDelta defines the number of blocks before the
```
excellent documentation 👍

joostjager commented:
@halseth ptal
